presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java

```diff
@@ -0,0 +1,163 @@
+                .collect(Collectors.toSet());
+
+        if (singletonInputs.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // exclude any inputs to TRY expression. Inlining them would potentially
+        // change the semantics of those expressions
+        Set<Symbol> tryArguments = parent.getAssignments()
+                .getExpressions().stream()
+                .flatMap(e -> extractTryArguments(e).stream())
+                .collect(Collectors.toSet());
+
+        Set<Symbol> candidates = Sets.difference(singletonInputs, tryArguments);
+
+        // see if any assignments for those inputs is *not* an identity projection
+        boolean allIdentities = child.getAssignments()
+                .entrySet().stream()
+                .filter(e -> candidates.contains(e.getKey()))
+                .allMatch(e -> e.getValue() instanceof SymbolReference &&
+                        e.getKey().getName().equals(((SymbolReference) e.getValue()).getName()));
+        if (allIdentities) {
+            return Optional.empty();
+        }
+
+        // inline the expressions
+        Assignments.Builder parentAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> entry : parent.getAssignments().entrySet()) {
+            Expression inlined = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Object>()
+            {
+                @Override
+                public Expression rewriteSymbolReference(SymbolReference reference, Object context, ExpressionTreeRewriter treeRewriter)
+                {
+                    Symbol symbol = Symbol.from(reference);
+                    if (candidates.contains(symbol)) {
+                        return child.getAssignments().get(symbol);
+                    }
+
+                    return reference;
+                }
+            }, entry.getValue());
+
+            parentAssignments.put(entry.getKey(), inlined);
+        }
+
+        // derive identity assignments for the inputs of expressions that were inlined
+        Set<Symbol> inputs = child.getAssignments()
+                .entrySet().stream()
+                .filter(e -> candidates.contains(e.getKey()))
+                .map(Map.Entry::getValue)
+                .flatMap(e -> DependencyExtractor.extractAll(e).stream())
+                .collect(Collectors.toSet());
+
+        Assignments.Builder childAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> assignment : child.getAssignments().entrySet()) {
+            if (!candidates.contains(assignment.getKey())) {
+                childAssignments.put(assignment.getKey(), assignment.getValue());
+            }
+        }
+        for (Symbol input : inputs) {
+            childAssignments.put(input, input.toSymbolReference());
+        }
+
+        return Optional.of(
+                new ProjectNode(
+                        parent.getId(),
+                        new ProjectNode(
```

Comment:
I see your point, your example seems to be CBO related.

Though, `ProjectNode`s are used very commonly. Typical plan (before optimization) has a lot them. Many (if not most) of them are very simple which could be merged without introducing any performance regression.

For example things like:
```
 project (a, b, 1 as c)
  - project(a, a * 2 as b)
    - scan(a)
```
could be simple merged to
```
project (a, a * 2 as b, 1 as c)
 - scan(a)
```

Above cases (or identity projections) could be handled by separate rule, but IMO it should be provided so we could say we have covered legacy `MergeProjections` in decent way.
I do not expect any performance boost because of that, but at least plan would be less polluted.